pkg/scd/constraints_handler.go

```diff
@@ -2,29 +2,350 @@ package scd
+	"github.com/interuss/dss/pkg/scd/repos"
 )
 
 // DeleteConstraintReference deletes a single constraint ref for a given ID at
 // the specified version.
 func (a *Server) DeleteConstraintReference(ctx context.Context, req *scdpb.DeleteConstraintReferenceRequest) (*scdpb.ChangeConstraintReferenceResponse, error) {
-	return nil, dsserr.BadRequest("not yet implemented")
+	// Retrieve Constraint ID
+	idString := req.GetEntityuuid()
+	if idString == "" {
+		return nil, dsserr.BadRequest("missing Constraint ID")
+	}
+	id := scdmodels.ID(idString)
+
+	// Retrieve ID of client making call
+	owner, ok := auth.OwnerFromContext(ctx)
+	if !ok {
+		return nil, dsserr.PermissionDenied("missing owner from context")
+	}
+
+	var response *scdpb.ChangeConstraintReferenceResponse
+	action := func(ctx context.Context, r repos.Repository) (err error) {
+		// Make sure deletion request is valid
+		old, err := r.GetConstraint(ctx, id)
+		switch {
+		case err == sql.ErrNoRows:
+			return dsserr.NotFound(id.String())
+		case err != nil:
+			return err
+		case old.Owner != owner:
+			return dsserr.PermissionDenied(fmt.Sprintf("constraint is owned by %s", old.Owner))
+		}
+
+		// Identify cells possibly containing Subscriptions to be notified
+		cells, err := r.GetConstraintCells(ctx, id)
+		if err != nil {
+			return err
+		}
+
+		// Find Subscriptions that may overlap the Constraint's Volume4D
+		allsubs, err := r.SearchSubscriptions(ctx, &dssmodels.Volume4D{
+			StartTime: old.StartTime,
+			EndTime:   old.EndTime,
+			SpatialVolume: &dssmodels.Volume3D{
+				AltitudeHi: old.AltitudeUpper,
+				AltitudeLo: old.AltitudeLower,
+				Footprint: dssmodels.GeometryFunc(func() (s2.CellUnion, error) {
+					return cells, nil
+				}),
+			}})
+		if err != nil {
+			return err
+		}
+
+		// Limit Subscription notifications to only those interested in Constraints
+		subs := make([]*scdmodels.Subscription, len(allsubs))
+		for _, sub := range allsubs {
+			if sub.NotifyForConstraints {
+				subs = append(subs, sub)
+			}
+		}
+
+		// Delete Constraint in Store
+		err = r.DeleteConstraint(ctx, id)
+		if err != nil {
+			return err
+		}
+
+		// Convert deleted Constraint to proto
+		constraintProto, err := old.ToProto()
+		if err != nil {
+			return dsserr.Internal(err.Error())
+		}
+
+		// Return response to client
+		response = &scdpb.ChangeConstraintReferenceResponse{
+			ConstraintReference: constraintProto,
+			Subscribers:         makeSubscribersToNotify(subs),
+		}
+
+		return nil
+	}
+
+	err := a.Store.Transact(ctx, action)
+	if err != nil {
+		return nil, err
+	}
+
+	return response, nil
 }
 
 // GetConstraintReference returns a single constraint ref for the given ID.
 func (a *Server) GetConstraintReference(ctx context.Context, req *scdpb.GetConstraintReferenceRequest) (*scdpb.GetConstraintReferenceResponse, error) {
-	return nil, dsserr.BadRequest("not yet implemented")
+	id := scdmodels.ID(req.GetEntityuuid())
+	if id.Empty() {
```

Comment:
Same nit for folding applies here.

Reply:
I'm not sure I follow on this one; seems like this is already folded per suggestion, with the only difference being `id.Empty()` instead of `len(id) == 0`

Reply:
instead of all of that, could we create the model, and the call `err := constraint.Validate()`

Reply:
@steeling could you elaborate?  This is a request to get a constraint, so we don't already have the constraint to `.Validate()` at this point.

Reply:
ya this comment should be on the Put/Update api